Generate server block configuration as well as location, fixes #5

samdjstevens opened the pull request:
This PR gives the plugin the ability to generate server block location as well as location blocks, with the option to set one to `nil` to not generate - fixes #5.

Commit:
Generate server block configuration as well as location, make both optional

sjamaan commented on README.md:
```diff
@@ -25,7 +25,7 @@ Install the plugin as usual:
 First, install NGINX and create a configuration as usual.  Then, in
 the `server` configuration block for the host you want to use for
-proxying, simply put `include "vagrant-proxy-config";` in the file.
+proxying, simply put `include "vagrant-proxy-config-locations";` in the file.
```
This is a pretty big backwards incompatible breaking change.

The name is clearer though, so I don't really mind making the next release a new major version.

sjamaan commented on README.md:
```diff
@@ -46,9 +46,20 @@ proxy to port 80 on the virtual machine with a `config.vm.hostname`
+#### Server Blocks
+
+The plugin also writes `server` block configuration for the enabled VMs so that they can be
+accessed directly via their hostname (as long as the hostname resolves to the host machine's I.P address).
```
Stupid nitpick: Just make it "IP address". The period is weird here, even moreso because it's only following the I but not the P.

sjamaan commented on lib/vagrant-reverse-proxy/action/write_nginx_config.rb:
```diff
@@ -101,6 +128,37 @@ def server_block(machine)
+          end
+          ip = get_ip_address(machine)
+          vhosts.collect do |path, vhost|
+            # Rewrites are matches literally by nginx, which means
```
This is problematic: The `path` is not used.  In my case, I have a config that looks like this:

```
config.reverse_proxy.vhosts = {
   "nl.dr.test" => "nl.dr.test",
   "nl.dr.test_node" => {:host => "nl.dr.test", :port => 8013},
   "be.dr.test" => "nl.dr.test",
   "be.dr.test_node" => {:host => "nl.dr.test", :port => 9013},
}
```

The idea is that there are two versions of the application, a version for the Netherlands and one for Belgium. Both have a NodeJS websocket server running, on different ports.  You can reach the Dutch application via `http://localhost/nl.dr.test/` and the websocket of the Dutch application via `http://localhost/nl.dr.test_node/`.

Unfortunately, the server-based config now gets generated like this:

```
server {
    listen 80;
    listen 443 ssl;
    server_name nl.dr.test;
    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://10.10.10.70/;
    }
}

server {
    listen 80;
    listen 443 ssl;
    server_name nl.dr.test;
    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://10.10.10.70:8013/;
    }
}

server {
    listen 80;
    listen 443 ssl;
    server_name be.dr.test;
    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://10.10.10.70/;
    }
}

server {
    listen 80;
    listen 443 ssl;
    server_name be.dr.test;
    location / {
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        proxy_set_header X-Forwarded-Host $host;
        proxy_set_header X-Forwarded-Port $server_port;
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header X-Forwarded-Proto $scheme;
        proxy_pass http://10.10.10.70:9013/;
    }
}
```

As you can see, it generates two server blocks per version that are identical except for the `proxy_pass` rule.  This means that the websocket will be unreachable. If I had swapped the order of the two, the application would be unreachable because it picks the first match.

sjamaan commented on lib/vagrant-reverse-proxy/action/write_nginx_config.rb:
```diff
@@ -101,6 +128,37 @@ def server_block(machine)
+            port_suffix = vhost[:port] == 80 ? '' : ":#{vhost[:port]}"
+            <<EOF
+server {
+    listen 80;
```
This caused the following error:

```
Dec 08 09:18:33 windu nginx[2800]: nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
Dec 08 09:18:33 windu nginx[2800]: nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
Dec 08 09:18:34 windu nginx[2800]: nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
Dec 08 09:18:34 windu nginx[2800]: nginx: [emerg] still could not bind()
Dec 08 09:18:34 windu systemd[1]: nginx.service: control process exited, code=exited status=1
Dec 08 09:18:34 windu systemd[1]: Failed to start A high performance web server and a reverse proxy server.
-- Subject: Unit nginx.service has failed
```

This happens because I have the following in my `default` site:

```
        listen [::]:80 default ipv6only=off;
```

Changing the generated line from `listen 80` to `listen [::]:80` did the trick for me. This probably has to do with another stupid literal match: Nginx isn't clever enough to understand that both these statements have the effect of binding to the same port on ipv4.

Can you check if this works for you as well? I'm not sure how robust this is, though. I suppose if you have `listen 80` in the default site, this might break as well (because it doesn't match literally, or it may start listening on ipv6 only?)

sjamaan commented on lib/vagrant-reverse-proxy/action/write_nginx_config.rb:
```diff
@@ -15,21 +15,49 @@ def call(env)
-          # Determine temp file and target file
-          nginx_config_file = env[:machine].config.reverse_proxy.nginx_config_file || '/etc/nginx/vagrant-proxy-config'
+          # Determine the filepaths for the locations config file. Give a default if not set.
+          if env[:machine].config.reverse_proxy.nginx_locations_config_file == VagrantPlugins::ReverseProxy::Plugin::Config::UNSET_VALUE
```
Checking against `VagrantPlugins::ReverseProxy::Plugin::Config::UNSET_VALUE` here is a bit iffy. It makes more sense to instead just set the default in the `finalize!` method when you encounter the `UNSET_VALUE`. That way you can remove all 6 lines here.

sjamaan commented on lib/vagrant-reverse-proxy/config.rb:
```diff
@@ -31,9 +32,6 @@ def finalize!
             end
           end
-          if @nginx_config_file == UNSET_VALUE
-            @nginx_config_file = nil
```
See comment above: I would just change this to:

```
if @nginx_location_config_file == UNSET_VALUE
  @nginx_location_config_file = '/etc/nginx/vagrant-proxy-config-locations'
end
if @nginx_servers_config_file == UNSET_VALUE
  @nginx_servers_config_file = '/etc/nginx/vagrant-proxy-config-servers'
end
```

The idea of `finalize!` is to set defaults and normalize the configuration options. This cleans up the logic in the rest of the plugin because you only have to deal with all the nasty bits here.

sjamaan reviewed (changes requested):
Thanks for the PR! Most of it looks good, but I have a few nitpicks and two issues with the servers config file: it doesn't work due to binding to the wrong port, and there's no way to dispatch to different ports in the target VM.

Commit:
PR fixes

samdjstevens commented:
Hey, 

Thanks for the feedback - 

1) Next version being a new major sounds good to keep the new default name of the locations config file. I've also took this opportunity to slightly alter the host NGINX configuration needed in the README (explained below).

2) Changed to IP address

3) Good idea with setting the defaults in the finalize! method - changed to do this.

4) I've fixed this now to use the `path` as the server name, and the `vhost[:host]` as the `Host` header value. I had a go with your example settings and it now seems to work as intended.

5) I think the most robust way to fix this port binding error is to specify both forms in both the generated server blocks, and the default host block. What do you think?

sjamaan commented on lib/vagrant-reverse-proxy/config.rb:
```diff
@@ -32,6 +32,12 @@ def finalize!
               end
             end
           end
+          if @nginx_location_config_file == UNSET_VALUE
```
This is a typo (it should be `@nginx_locations_config_file`) which causes this if to always fail (undefined instance variables are always `nil`), and therefore we get a crash when the `UNSET_VALUE` is used as a filename when writing the config!

```
in `dirname': no implicit conversion of Object into String (TypeError)
```

sjamaan commented on lib/vagrant-reverse-proxy/action/write_nginx_config.rb:
```diff
@@ -15,19 +15,9 @@ def call(env)
-          else
-            nginx_servers_config_file = env[:machine].config.reverse_proxy.nginx_servers_config_file
-          end
+          # Get the filepaths for the config files.
```
This still looks a bit noisy to me.  I was thinking, we could just unconditionally do this:

```
generate_config_file(env, env[:machine].config.reverse_proxy.nginx_locations_config_file, location_block(env[:machine]))
generate_config_file(env, env[:machine].config.reverse_proxy.nginx_servers_config_file, location_block(env[:machine]))
```

If this is a bit too long or repetitive, a nice way to clean this up could be to first say `rp_config = env[:machine].config.reverse_proxy`, then use that everywhere we access a config setting.

Then, let `generate_config_file` deal with undefined paths like so:

```
return unless file_to_write_to
```

As an aside, I think `target_file` would be a less clumsy name than `file_to_write_to`.

What do you think?

sjamaan commented on lib/vagrant-reverse-proxy/action/write_nginx_config.rb:
```diff
@@ -143,10 +133,13 @@ def server_block(machine)
             <<EOF
 server {
     listen 80;
+    listen [::]:80;
```
Simply adding this didn't make the problem go away.

When I changed my default site's config to match this, it worked:

```
        listen 80 default;
        listen [::]:80 default;
```

I think this is acceptable, but it means the README needs to be changed, because with `ipv6only=off` behind the `listen [::]80 default` (as the README now says), it **didn't** work.

sjamaan commented on lib/vagrant-reverse-proxy/action/write_nginx_config.rb:
```diff
@@ -143,10 +133,13 @@ def server_block(machine)
+    server_name #{path};
     location / {
-        proxy_set_header Host $host;
+        proxy_set_header Host #{vhost[:host]};
```
Excellent, this did the trick!

sjamaan reviewed (changes requested):
Thanks for responding so quickly once again. I noticed one bug due to a typo, one problem with the `listen` lines in the README and I think the code will be more compact if we move the `nil` check from the `call` method  to `generate_config_file`.

Commit:
some more PR fixes

samdjstevens commented:
Thanks for the feedback again - think I've got them all sorted!

sjamaan commented:
This is great work. Many thanks for your patience!

I have tagged 0.4.0 and released it, so you should be able to install it from Rubygems now, using `vagrant plugin update vagrant-reverse-proxy`.

samdjstevens commented:
Excellent, cheers! Thanks for guiding me through it haha.
